### PR TITLE
Sort events 5148 online by date offline by location

### LINKED ISSUE #5148: [View my events] The upcoming events are displayed in random order on My events page
**Environment:** MacOS 12.0.1, GoogleChrom 109.0.5414.87
**Reproducible:** always, rarely, sometimes etc.
**Build found:** 31.01.2023
E.g. "commit 147fa44"

**Preconditions**

**Steps to reproduce**
1. Go to https://www.greencity.social/#/greenCity and log in
2. Go to Events page and join 2 online and 2 offline events
3. Go to My Events and view added events

**Actual result**
The online and offline events are displayed  in random order

**Expected result**
1. The online events are filtered by date, so that later event has higher position in list
2. The offline events are filtered by place closer to the User's IP location, so that closer events have higher position in the list

**Postconditions**

DELETE FROM table1
WHERE id=<item1_id>;"

**User story and test case links**
E.g.: "User story #3375 
[Test case](https://jira.softserve.academy/browse/GC-5018)"

**Labels to be added**
"Bug", Priority ("pri: "), Severity ("severity:"), Type ("UI, "Functional"), "API" (for back-end bugs).

### PR DESCRIPTION
**Pull request by HelenSotnik**
# GreenCity PR
[[View my events] The upcoming events are displayed in random order on My events page #5148](https://github.com/ita-social-projects/GreenCity/issues/5148)

## Summary Of Changes :
Sorting online events by date and offline by closeness to user geoposition if user shows his location, otherwise, by date. If no eventType is chosen all user events are sorted by date.

## Added methods in EventServiceImpl :
getComparatorByDistance
getComparatorByDates
calculateDistance (to calculate distance between event attender and event location)
getOfflineUserEventsSortedCloserToUserLocation (if filtration choice is online and event attender share his location)
getOfflineUserEventsSortedByDate (if filtration choice is online and even tattender share his location)
getOnlineUserEventsSortedByDate (if filtration choice is online)
sortUserEventsByEventType (method accept eventType and gives sorted List list)

## Changed methods in EventServiceImpl:
getAllUserEvents

## How to test
Created test for all new methods.
Test with local database are made in Postman with different scenarios, show correct endpoint data.
Necessary to add 3 Params to succeed in testing:
eventType( may be: online,offline, or empty string)
userLatitude(may be: any string representing number for example 50.42929, or empty string)
userLongitude(may be: any string representing number,for example 30.53806 or empty string)
image
image
image
image

### CODE REVIEW

File: service/src/main/java/greencity/service/EventServiceImpl.java [hunk 1]
```diff
@@ -137,14 +127,105 @@ public PageableAdvancedDto<EventDto> getAll(Pageable page, Principal principal)
...
     }
 
+    private List<Event> sortUserEventsByEventType(
+        String eventType, User attender, String userLatitude, String userLongitude) {
+        if (eventType.equalsIgnoreCase("ONLINE")) {
+            return getOnlineUserEventsSortedByDate(attender);
+        }
+
+        if (eventType.equalsIgnoreCase("OFFLINE")) {
+            if (!userLatitude.isBlank() && !userLongitude.isBlank()) {
```
Comment (VadymMakitra) [inline service/src/main/java/greencity/service/EventServiceImpl.java]:
```suggestion
        if (eventType.equalsIgnoreCase("OFFLINE") && !userLatitude.isBlank() && !userLongitude.isBlank()) {
```

File: service/src/main/java/greencity/service/EventServiceImpl.java [hunk 2]
```diff
@@ -137,14 +127,105 @@ public PageableAdvancedDto<EventDto> getAll(Pageable page, Principal principal)
...
+                return getOfflineUserEventsSortedByDate(attender);
+            }
+        }
+        return eventRepo.findAllByAttender(attender.getId()).stream().sorted(getComparatorByDates())
+            .collect(Collectors.toList());
+    }
+
+    private List<Event> getOnlineUserEventsSortedByDate(User attender) {
+        return eventRepo.findAllByAttender(attender.getId()).stream()
+            .filter(event -> event.getDates().get(event.getDates().size() - 1).getOnlineLink() != null)
```
Comment (VadymMakitra) [inline service/src/main/java/greencity/service/EventServiceImpl.java]: What are you trying to do here?

File: service/src/main/java/greencity/service/EventServiceImpl.java [hunk 3]
```diff
@@ -137,14 +127,105 @@ public PageableAdvancedDto<EventDto> getAll(Pageable page, Principal principal)
...
+            .sorted(getComparatorByDates())
+            .collect(Collectors.toList());
+    }
+
+    private List<Event> getOfflineUserEventsSortedByDate(User attender) {
+        return eventRepo.findAllByAttender(attender.getId()).stream()
+            .filter(event -> Objects.requireNonNull(event.getDates().get(event.getDates().size() - 1)
+                .getAddress()).getLatitude() != 0
+                && Objects.requireNonNull(event.getDates().get(event.getDates().size() - 1)
+                    .getAddress()).getLongitude() != 0)
```
Comment (VadymMakitra) [inline service/src/main/java/greencity/service/EventServiceImpl.java]: What are you trying to do here?

File: service/src/main/java/greencity/service/EventServiceImpl.java [hunk 4]
```diff
@@ -137,14 +127,105 @@ public PageableAdvancedDto<EventDto> getAll(Pageable page, Principal principal)
...
+            .compareTo(e1.getDates().get(e1.getDates().size() - 1).getStartDate());
+    }
+
+    private static double calculateDistance(double latitude1, double longitude1, double latitude2, double longitude2) {
+        double earthRadius = 6371.0;
+
+        double latitude1Rad = Math.toRadians(latitude1);
+        double longitude1Rad = Math.toRadians(longitude1);
+        double latitude2Rad = Math.toRadians(latitude2);
+        double longitude2Rad = Math.toRadians(longitude2);
```
Comment (VadymMakitra) [inline service/src/main/java/greencity/service/EventServiceImpl.java]: Please rename the variables

File: service/src/main/java/greencity/service/EventServiceImpl.java [hunk 5]
```diff
@@ -137,14 +127,105 @@ public PageableAdvancedDto<EventDto> getAll(Pageable page, Principal principal)
...
+        double distanceLatitude = latitude2Rad - latitude1Rad;
+        double distanceLongitude = longitude2Rad - longitude1Rad;
+
+        double a = Math.sin(distanceLatitude / 2) * Math.sin(distanceLatitude / 2)
```
Comment (VadymMakitra) [inline service/src/main/java/greencity/service/EventServiceImpl.java]: Please rename this

File: service/src/main/java/greencity/service/EventServiceImpl.java [hunk 6]
```diff
@@ -137,14 +127,105 @@ public PageableAdvancedDto<EventDto> getAll(Pageable page, Principal principal)
...
+            + Math.cos(latitude1Rad) * Math.cos(latitude2Rad) * Math.sin(distanceLongitude / 2)
+                * Math.sin(distanceLongitude / 2);
+        double c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
```
Comment (VadymMakitra) [inline service/src/main/java/greencity/service/EventServiceImpl.java]: Please rename this

**Review by VadymMakitra — COMMENTED** [review]
I didn't get what you want to sort in the stream method written by you.

File: service/src/main/java/greencity/service/EventServiceImpl.java (hunk 1 above)
Comment (HelenSotnik) [reply service/src/main/java/greencity/service/EventServiceImpl.java]: I have two cases to take into consideration when it comes about OFFLINE. First, when user share geo location and userLatitude and userLongitude come from frontend, the other,this params are empty string as user douesn't share geo location. So I desided to change this way:  
                                                                            
if (eventType.equalsIgnoreCase("OFFLINE")) {
            return (!userLatitude.isBlank() && !userLongitude.isBlank()) 
                    ?getOfflineUserEventsSortedCloserToUserLocation(attender, userLatitude, userLongitude) :
                    getOfflineUserEventsSortedByDate(attender);
        }

File: service/src/main/java/greencity/service/EventServiceImpl.java (hunk 2 above)
Comment (HelenSotnik) [reply service/src/main/java/greencity/service/EventServiceImpl.java]: I tried to filter only online events, and they are online if event online link is not null. But I have changed this part of code into proposed by you method (getEventType()) in entity class Event. And filter works using this method now.

File: service/src/main/java/greencity/service/EventServiceImpl.java (hunk 3 above)
Comment (HelenSotnik) [reply service/src/main/java/greencity/service/EventServiceImpl.java]: I tried to filter offline events, but I have changed this part of code using above mentioned method getEventType() in Event class, that returns enum(ONLINE, OFFLINE or ONLINE_OFFLINE).

File: service/src/main/java/greencity/service/EventServiceImpl.java (hunk 4 above)
Comment (HelenSotnik) [reply service/src/main/java/greencity/service/EventServiceImpl.java]: I have changed the implementation of current method, instead I added dependency in pom.xml org.locationtech.jts, in order to make method be more presize( the previous considered Earth as a perfect ellipce, which is approximated the calculations). Library added helps to make more specific calculations of distance beetween two points in the map.

File: service/src/main/java/greencity/service/EventServiceImpl.java [hunk 7]
```diff
@@ -137,14 +143,96 @@ public PageableAdvancedDto<EventDto> getAll(Pageable page, Principal principal)
...
     }
 
+    private List<Event> sortUserEventsByEventType(
+        String eventType, User attender, String userLatitude, String userLongitude) {
+        if (eventType.equalsIgnoreCase("ONLINE")) {
+            return getOnlineUserEventsSortedByDate(attender);
+        }
+
+        if (eventType.equalsIgnoreCase("OFFLINE")) {
+            return (!userLatitude.isBlank() && !userLongitude.isBlank())
```
Comment (VadymMakitra) [inline service/src/main/java/greencity/service/EventServiceImpl.java]:
```suggestion
            return (StringUtils.isNotBlank(userLatitude) && StringUtils.isNotBlank(userLongitude)
```

Comment (VadymMakitra) [reply service/src/main/java/greencity/service/EventServiceImpl.java]: potentially could produce NPE

File: service/src/main/java/greencity/service/EventServiceImpl.java [hunk 8]
```diff
@@ -137,14 +143,96 @@ public PageableAdvancedDto<EventDto> getAll(Pageable page, Principal principal)
...
+                ? getOfflineUserEventsSortedCloserToUserLocation(attender, userLatitude, userLongitude)
```
Comment (VadymMakitra) [inline service/src/main/java/greencity/service/EventServiceImpl.java]:
```suggestion
                ? getOfflineUserEventsSortedByUserLocation(attender, userLatitude, userLongitude)
```

File: service/src/main/java/greencity/service/EventServiceImpl.java [hunk 9]
```diff
@@ -137,14 +143,96 @@ public PageableAdvancedDto<EventDto> getAll(Pageable page, Principal principal)
...
+            .filter(this::isEventRelevant)
+            .collect(Collectors.toList());
+        List<Event> eventsPassed = getOfflineUserEventsSortedByDate(attender).stream()
+            .filter(event -> !isEventRelevant(event))
+            .collect(Collectors.toList());
+        eventsFurtherSorted.addAll(eventsPassed);
+        return eventsFurtherSorted;
+    }
+
+    private boolean isEventRelevant(Event event) {
```
Comment (VadymMakitra) [inline service/src/main/java/greencity/service/EventServiceImpl.java]: move this to the Event class

File: dao/src/test/java/greencity/entity/event/EventTest.java [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+
+        assertEquals(EventType.ONLINE, eventType);
+    }
+
+    @Test
+    public void testGetEventTypeOnlineOffline() {
+        Event event = ModelUtils.getOnlineOfflineEvent();
+        EventType eventType = event.getEventType();
+
+        assertEquals(EventType.ONLINE_OFFLINE, eventType);
```
Comment (VadymMakitra) [inline dao/src/test/java/greencity/entity/event/EventTest.java]: These are not proper tests.

File: dao/src/main/java/greencity/entity/event/Event.java [hunk 11]
```diff
@@ -81,12 +83,28 @@ public class Event {
...
     private List<EventGrade> eventGrades = new ArrayList<>();
+
+    /**
+     * Method for getting an event type.
+     *
+     * @return {@link EventType} instance.
+     * @author Olena Sotnik.
+     */
+    public EventType getEventType() {
+        if (dates.stream().allMatch(date -> Objects.nonNull(date.getOnlineLink())
```
Comment (VadymMakitra) [inline dao/src/main/java/greencity/entity/event/Event.java]: probably anyMatch

File: dao/src/main/java/greencity/entity/event/Event.java [hunk 12]
```diff
@@ -81,12 +83,28 @@ public class Event {
...
+            && Objects.nonNull(date.getAddress()))) {
+            return EventType.ONLINE_OFFLINE;
+        }
+        if (dates.stream().anyMatch(date -> Objects.nonNull(date.getOnlineLink()))) {
```
Comment (VadymMakitra) [inline dao/src/main/java/greencity/entity/event/Event.java]: allMatch

File: service/src/main/java/greencity/service/EventServiceImpl.java (hunk 7 above)
Comment (HelenSotnik) [reply service/src/main/java/greencity/service/EventServiceImpl.java]: I've changed as proposed.

File: service/src/main/java/greencity/service/EventServiceImpl.java (hunk 8 above)
Comment (HelenSotnik) [reply service/src/main/java/greencity/service/EventServiceImpl.java]: I have changed method name as proposed above.

File: service/src/main/java/greencity/service/EventServiceImpl.java (hunk 9 above)
Comment (HelenSotnik) [reply service/src/main/java/greencity/service/EventServiceImpl.java]: I have moved to Event class method and named it isRelevant()

File: dao/src/main/java/greencity/entity/event/Event.java (hunk 11 above)
Comment (HelenSotnik) [reply dao/src/main/java/greencity/entity/event/Event.java]: I have changed as proposed.

File: dao/src/main/java/greencity/entity/event/Event.java (hunk 12 above)
Comment (HelenSotnik) [reply dao/src/main/java/greencity/entity/event/Event.java]: I have changed as proposed.

**Review by VadymMakitra — APPROVED** [review]

**Review by ABbondar — APPROVED** [review]

**Review by Yurii-Midianyi — APPROVED** [review]

**Pull request merged**